### PR TITLE
Meta Package Manager 5.13.0 (new formula)

### PR DESCRIPTION
**Pull request by kdeldycke**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

This is a new formula for Meta Package Manager, a multi-platform pure-Python project, which provides the `mpm` CLI, a wrapper around all kind of package managers.

It is an actively maintain project and starts to gain traction. A user [already made its own formula a while back](https://github.com/Hasnep/homebrew-tap/blob/main/Formula/meta-package-manager.rb), which prove the interested of having it available via `brew`. So it is natural to propose it for integration upstream via this PR.

### CODE REVIEW

**Comment by kdeldycke** [comment]
All tests and audits pass perfectly on my machine, but I got a strange issue with `brew test`:
```shell-session
$ brew test meta-package-manager
Error: meta-package-manager is missing test dependencies: sqlite xz python3
```

I'm not certain this is a serious problem and probably related to my own environment. I'm proposing the PR anyway so I could double-check here if the CI is pointing to the same issue.

File: Formula/meta-package-manager.rb [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+    sha256 "341595a488e3e01a85a9d8911d8912fd922ede5fecc4dce437eb4b6c8d037e56"
+  end
+
+  resource "zipp" do
+    url "https://files.pythonhosted.org/packages/ab/47/b47d02b741e0aa6f998fc80457d3dfc05cb7732ef480597c2971cbc79260/zipp-3.14.0.tar.gz"
+    sha256 "9e5421e176ef5ab4c0ad896624e87a7b2f07aca746c9b2aa305952800cb8eecb"
+  end
+
+  def install
+    virtualenv_create(libexec, "python3")
```
Comment (dawidd6) [inline Formula/meta-package-manager.rb:408]:
```suggestion
```

I think this line is unnecessary, `virtualenv_install_with_resources` already does that.

Comment (kdeldycke) [reply Formula/meta-package-manager.rb]: Thanks @dawidd6 for the feedback. Just removed it. Will see how the CI react.

**Comment by kdeldycke** [comment]
All tests are passing and successful. The code review changes have been applied. This PR is ready to be merged upstream.

**Comment by kdeldycke** [comment]
Credit were credit is due, the initial packaging of Meta Package Manager for Homebrew was proposed by @Hasnep last year at: https://github.com/Hasnep/homebrew-tap

File: Formula/meta-package-manager.rb [hunk 2]
```diff
@@ -0,0 +1,417 @@
+class MetaPackageManager < Formula
+  include Language::Python::Virtualenv
+
+  desc "🎁 wraps all package managers with a unifying CLI"
```
Comment (ZhongRuoyu) [inline Formula/meta-package-manager.rb:4]: Though less fun, this is probably easier to understand:

```suggestion
  desc "Wrapper around all package managers with a unifying CLI"
```

File: Formula/meta-package-manager.rb [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+  homepage "https://github.com/kdeldycke/meta-package-manager"
```
Comment (ZhongRuoyu) [inline Formula/meta-package-manager.rb:5]: Following what's shown on the [GitHub repo page](https://github.com/kdeldycke/meta-package-manager).

```suggestion
  homepage "https://kdeldycke.github.io/meta-package-manager/"
```

File: Formula/meta-package-manager.rb [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+  url "https://files.pythonhosted.org/packages/f5/28/5235fca899ead8059fbb35dd4bd786c1dbd4e1d0d65ae2d136540681a60c/meta_package_manager-5.13.0.tar.gz"
+  sha256 "8452718bbcec58dac5fd27691b1be621fcc0d25eb54cb3d3a7db89cfd7f2b1d3"
+  license "GPL-2.0-only"
```
Comment (ZhongRuoyu) [inline Formula/meta-package-manager.rb:8]: This is actually [`GPL-2.0-or-later`](https://github.com/kdeldycke/meta-package-manager/blob/main/license#L299).

```suggestion
  license "GPL-2.0-or-later"
```

File: Formula/meta-package-manager.rb [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+    url "https://files.pythonhosted.org/packages/94/71/a8ee96d1fd95ca04a0d2e2d9c4081dac4c2d2b12f7ddb899c8cb9bfd1532/alabaster-0.7.13.tar.gz"
+    sha256 "a27a4a084d5e690e16e01e03ad2b2e552c61a65469419b907243193de1a84ae2"
+  end
+
+  resource "attrs" do
+    url "https://files.pythonhosted.org/packages/21/31/3f468da74c7de4fcf9b25591e682856389b3400b4b62f201e65f15ea3e07/attrs-22.2.0.tar.gz"
+    sha256 "c9227bfc2f01993c03f68db37d1d15c9690188323c067c641f1a35ca58185f99"
+  end
+
+  resource "babel" do
```
Comment (ZhongRuoyu) [inline Formula/meta-package-manager.rb:22]:
```suggestion
  resource "Babel" do
```

File: Formula/meta-package-manager.rb [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+
+  def install
+    virtualenv_install_with_resources
+  end
+
+  test do
+    system bin/"mpm", "--version"
+    system bin/"mpm", "--help"
+    system bin/"mpm", "--all-managers", "managers"
+    system bin/"mpm", "installed"
```
Comment (ZhongRuoyu) [inline Formula/meta-package-manager.rb:415]: Can the output be checked? That'd be more robust than merely executing the commands.

File: Formula/meta-package-manager.rb (hunk 5 above)
Comment (ZhongRuoyu) [reply Formula/meta-package-manager.rb:22]: Also, several PyPI resources can be converted to formula dependencies. For example, `docutils` and `pyyaml`. Can I request you to identify all, make them formula dependencies, and add an entry to [`pypi_formula_mappings.json`](https://github.com/Homebrew/homebrew-core/blob/master/pypi_formula_mappings.json)? The PyPI package names should be listed in `exclude_packages`.

File: Formula/meta-package-manager.rb (hunk 2 above)
Comment (kdeldycke) [reply Formula/meta-package-manager.rb]: Fair enough! I understand it's good to not be too fancy here.

File: Formula/meta-package-manager.rb (hunk 3 above)
Comment (kdeldycke) [reply Formula/meta-package-manager.rb:5]: Good to me! Let's merge this change!

File: Formula/meta-package-manager.rb (hunk 4 above)
Comment (kdeldycke) [reply Formula/meta-package-manager.rb:8]: Good point. The irony here is I'm the one who choose that license first! 😅

File: Formula/meta-package-manager.rb (hunk 5 above)
Comment (kdeldycke) [reply Formula/meta-package-manager.rb:22]: Oh OK. I see; the policy here is to rely as much as possible on formula already managed by `brew`. I'll make the necessary changes then.

File: Formula/meta-package-manager.rb (hunk 6 above)
Comment (kdeldycke) [reply Formula/meta-package-manager.rb:415]: Yes. I guess I can parse the output of these commands. I'll see what I can do here.

File: Formula/meta-package-manager.rb (hunk 5 above)
Comment (kdeldycke) [reply Formula/meta-package-manager.rb]: Just fixed the case of Python dependencies, and added direct dependencies on other Formula, namely: `docutils`, `pygments`, `python-tabulate`, `python-typing-extensions` and `pyyaml`.

`pypi_formula_mappings.json` has also been updated to reflect these changes.

File: Formula/meta-package-manager.rb [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+  def install
+    virtualenv_install_with_resources
+  end
+
+  test do
+    # Check mpm is reporting the correct version of itself.
+    assert_match(
+      /^mpm, version #{version}\n\{.*\}\n/,
+      shell_output("#{bin}/mpm --version")
+    )
```
Comment (ZhongRuoyu) [inline Formula/meta-package-manager.rb:396]: Maybe this:

```suggestion
    assert_match /^mpm, version #{version}\n\{.*\}\n/,
                 shell_output("#{bin}/mpm --version")
```

Comment (kdeldycke) [reply Formula/meta-package-manager.rb]: Thanks for the suggestion but no, use of regex requires parenthesis:

```shell-session
$ brew test-bot --only-tap-syntax
(...)
  /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/meta-package-manager.rb:393:18: W: [Correctable] Lint/AmbiguousRegexpLiteral: Ambiguous regexp literal. Parenthesize the method arguments if it's surely a regexp literal, or add a whitespace to the right of the / if it should be a division.
      assert_match /^mpm, version #{version}\n\{.*\}\n/,
```

See: https://github.com/Homebrew/homebrew-core/actions/runs/4652199961/jobs/8232262425?pr=127591#step:4:22

Comment (ZhongRuoyu) [reply Formula/meta-package-manager.rb]: Oh yea, sorry. On that note, is this regex really necessary, though? `mpm, version #{version}` actually looks fine. Same for the usage text.

As for the table outputs below, perhaps [`--output-format`](https://kdeldycke.github.io/meta-package-manager/cli-parameters.html#cmdoption-mpm-o) would help to simplify. The regexes look a bit more complicated than they should be.

Comment (kdeldycke) [reply Formula/meta-package-manager.rb]: Oh good ideas! I just pushed a simplified version of the tests. I did not realized `assert_match` allowed for partial matching, so a plain string of a subset of the expected output is enough.

File: Formula/meta-package-manager.rb (hunk 6 above)
Comment (kdeldycke) [reply Formula/meta-package-manager.rb]: Just pushed a little test suite associated with all these commands.

**Comment by kdeldycke** [comment]
All tests are now passing successfully and all comments pointed out during the code review have been addressed. This PR is ready to be merged upstream.

**Review by ZhongRuoyu — APPROVED** [review]
Looks good to me. Thanks @kdeldycke!

**Comment by kdeldycke** [comment]
> Looks good to me. Thanks @kdeldycke!

Thanks a lot @ZhongRuoyu for the detailed review! 🤗

**Review by chenrui333 — APPROVED** [review]

**Comment by kdeldycke** [comment]
🥳 I can confirm the package is available publicly and is working as expected:

```shell-session
$ brew update
Updated 1 tap (homebrew/core).
==> New Formulae
meta-package-manager
```

```shell-session
$ brew info meta-package-manager
==> meta-package-manager: stable 5.13.0 (bottled)
Wrapper around all package managers with a unifying CLI
https://kdeldycke.github.io/meta-package-manager/
Not installed
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/meta-package-manager.rb
License: GPL-2.0-or-later
==> Dependencies
Required: docutils ✘, pygments ✘, python-tabulate ✘, python-typing-extensions ✘, python@3.11 ✔, pyyaml ✘
```

```shell-session
$ brew install meta-package-manager
==> Fetching dependencies for meta-package-manager: docutils, pygments, python-tabulate, python-typing-extensions, libyaml and pyyaml
==> Fetching docutils
==> Downloading https://ghcr.io/v2/homebrew/core/docutils/manifests/0.19_1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/docutils/blobs/sha256:80e64efdb51243a4399e19e4dbf5efe01b855a7cad82ca79c0c6026042f7a00c
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:80e64efdb51243a4399e19e4dbf5efe01b855a7cad82ca79c0c6026042f7a00c?se=2023-04-10T14%3A45%3A00Z&sig=UBfz0fxzJxFHbE1B1phxCpaWPwHzlw
######################################################################## 100.0%
==> Fetching pygments
==> Downloading https://ghcr.io/v2/homebrew/core/pygments/manifests/2.14.0
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/pygments/blobs/sha256:d4b2214521e36150b21673e1e3b3e627a22919a98274bebeaa1fb8daa0ec086d
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:d4b2214521e36150b21673e1e3b3e627a22919a98274bebeaa1fb8daa0ec086d?se=2023-04-10T14%3A45%3A00Z&sig=ztdSKGfbQcU4qzJ7AhUaTvxDusqjxG
######################################################################## 100.0%
==> Fetching python-tabulate
==> Downloading https://ghcr.io/v2/homebrew/core/python-tabulate/manifests/0.9.0-1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/python-tabulate/blobs/sha256:2f4936777a24b5061f87e299884aef95136b4d252b923ae305d9bbe73e2dad28
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:2f4936777a24b5061f87e299884aef95136b4d252b923ae305d9bbe73e2dad28?se=2023-04-10T14%3A45%3A00Z&sig=cvM3s2GbkjiOnJXNaLqjjf8YRnglNQ
######################################################################## 100.0%
==> Fetching python-typing-extensions
==> Downloading https://ghcr.io/v2/homebrew/core/python-typing-extensions/manifests/4.5.0
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/python-typing-extensions/blobs/sha256:74d44da39b2998a115f06b2d345e4495eb018c031610a0102636b182c09e0949
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:74d44da39b2998a115f06b2d345e4495eb018c031610a0102636b182c09e0949?se=2023-04-10T14%3A45%3A00Z&sig=AOagfggB7oMsqTC%2Fnqxtgwp3vFAP
######################################################################## 100.0%
==> Fetching libyaml
==> Downloading https://ghcr.io/v2/homebrew/core/libyaml/manifests/0.2.5
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/libyaml/blobs/sha256:11239e8f5066c6d0d0718208d4eab518da00c7289f33c9c76c0a09ba5c0417c9
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:11239e8f5066c6d0d0718208d4eab518da00c7289f33c9c76c0a09ba5c0417c9?se=2023-04-10T14%3A50%3A00Z&sig=BSBBKfFWM7uMyO3Yor4v2uOP90La06
######################################################################## 100.0%
==> Fetching pyyaml
==> Downloading https://ghcr.io/v2/homebrew/core/pyyaml/manifests/6.0_1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/pyyaml/blobs/sha256:9cdc9ef202bb642be23ec38c05c0ab6ce9b08d09ba5d954fb6440c946aa31e56
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:9cdc9ef202bb642be23ec38c05c0ab6ce9b08d09ba5d954fb6440c946aa31e56?se=2023-04-10T14%3A50%3A00Z&sig=4PJpEJAmlg2Wi61wMssvyyb%2FAxZ0
######################################################################## 100.0%
==> Fetching meta-package-manager
==> Downloading https://ghcr.io/v2/homebrew/core/meta-package-manager/manifests/5.13.0
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/meta-package-manager/blobs/sha256:148c4e61dc29d373535caf5c1bf21f06a07cb1cd3aadda9e1a5f434b1938ab6d
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:148c4e61dc29d373535caf5c1bf21f06a07cb1cd3aadda9e1a5f434b1938ab6d?se=2023-04-10T14%3A50%3A00Z&sig=9xZeU2LIIZMNsD0kvIppLbuXId1Gtv
######################################################################## 100.0%
==> Installing dependencies for meta-package-manager: docutils, pygments, python-tabulate, python-typing-extensions, libyaml and pyyaml
==> Installing meta-package-manager dependency: docutils
==> Pouring docutils--0.19_1.arm64_ventura.bottle.tar.gz
🍺  /opt/homebrew/Cellar/docutils/0.19_1: 231 files, 2MB
==> Installing meta-package-manager dependency: pygments
==> Pouring pygments--2.14.0.all.bottle.tar.gz
Error: The `brew link` step did not complete successfully
The formula built, but is not symlinked into /opt/homebrew
Could not symlink bin/pygmentize
Target /opt/homebrew/bin/pygmentize
already exists. You may want to remove it:
  rm '/opt/homebrew/bin/pygmentize'

To force the link and overwrite all conflicting files:
  brew link --overwrite pygments

To list all files that would be deleted:
  brew link --overwrite --dry-run pygments

Possible conflicting files are:
/opt/homebrew/bin/pygmentize
==> Summary
🍺  /opt/homebrew/Cellar/pygments/2.14.0: 624 files, 8.1MB
==> Installing meta-package-manager dependency: python-tabulate
==> Pouring python-tabulate--0.9.0.arm64_ventura.bottle.1.tar.gz
🍺  /opt/homebrew/Cellar/python-tabulate/0.9.0: 29 files, 298.6KB
==> Installing meta-package-manager dependency: python-typing-extensions
==> Pouring python-typing-extensions--4.5.0.arm64_ventura.bottle.tar.gz
🍺  /opt/homebrew/Cellar/python-typing-extensions/4.5.0: 21 files, 236.2KB
==> Installing meta-package-manager dependency: libyaml
==> Pouring libyaml--0.2.5.arm64_ventura.bottle.tar.gz
🍺  /opt/homebrew/Cellar/libyaml/0.2.5: 10 files, 351.3KB
==> Installing meta-package-manager dependency: pyyaml
==> Pouring pyyaml--6.0_1.arm64_ventura.bottle.tar.gz
Error: The `brew link` step did not complete successfully
The formula built, but is not symlinked into /opt/homebrew
Could not symlink lib/python3.10/site-packages/_yaml/__init__.py
Target /opt/homebrew/lib/python3.10/site-packages/_yaml/__init__.py
already exists. You may want to remove it:
  rm '/opt/homebrew/lib/python3.10/site-packages/_yaml/__init__.py'

To force the link and overwrite all conflicting files:
  brew link --overwrite pyyaml

To list all files that would be deleted:
  brew link --overwrite --dry-run pyyaml

Possible conflicting files are:
/opt/homebrew/lib/python3.10/site-packages/_yaml/__init__.py
/opt/homebrew/lib/python3.10/site-packages/yaml/__init__.py
/opt/homebrew/lib/python3.10/site-packages/yaml/_yaml.cpython-310-darwin.so
/opt/homebrew/lib/python3.10/site-packages/yaml/composer.py
/opt/homebrew/lib/python3.10/site-packages/yaml/constructor.py
/opt/homebrew/lib/python3.10/site-packages/yaml/cyaml.py
/opt/homebrew/lib/python3.10/site-packages/yaml/dumper.py
/opt/homebrew/lib/python3.10/site-packages/yaml/emitter.py
/opt/homebrew/lib/python3.10/site-packages/yaml/error.py
/opt/homebrew/lib/python3.10/site-packages/yaml/events.py
/opt/homebrew/lib/python3.10/site-packages/yaml/loader.py
/opt/homebrew/lib/python3.10/site-packages/yaml/nodes.py
/opt/homebrew/lib/python3.10/site-packages/yaml/parser.py
/opt/homebrew/lib/python3.10/site-packages/yaml/reader.py
/opt/homebrew/lib/python3.10/site-packages/yaml/representer.py
/opt/homebrew/lib/python3.10/site-packages/yaml/resolver.py
/opt/homebrew/lib/python3.10/site-packages/yaml/scanner.py
/opt/homebrew/lib/python3.10/site-packages/yaml/serializer.py
/opt/homebrew/lib/python3.10/site-packages/yaml/tokens.py
/opt/homebrew/lib/python3.9/site-packages/_yaml/__init__.py
/opt/homebrew/lib/python3.9/site-packages/yaml/__init__.py
/opt/homebrew/lib/python3.9/site-packages/yaml/_yaml.cpython-39-darwin.so
/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py
/opt/homebrew/lib/python3.9/site-packages/yaml/constructor.py
/opt/homebrew/lib/python3.9/site-packages/yaml/cyaml.py
/opt/homebrew/lib/python3.9/site-packages/yaml/dumper.py
/opt/homebrew/lib/python3.9/site-packages/yaml/emitter.py
/opt/homebrew/lib/python3.9/site-packages/yaml/error.py
/opt/homebrew/lib/python3.9/site-packages/yaml/events.py
/opt/homebrew/lib/python3.9/site-packages/yaml/loader.py
/opt/homebrew/lib/python3.9/site-packages/yaml/nodes.py
/opt/homebrew/lib/python3.9/site-packages/yaml/parser.py
/opt/homebrew/lib/python3.9/site-packages/yaml/reader.py
/opt/homebrew/lib/python3.9/site-packages/yaml/representer.py
/opt/homebrew/lib/python3.9/site-packages/yaml/resolver.py
/opt/homebrew/lib/python3.9/site-packages/yaml/scanner.py
/opt/homebrew/lib/python3.9/site-packages/yaml/serializer.py
/opt/homebrew/lib/python3.9/site-packages/yaml/tokens.py
==> Summary
🍺  /opt/homebrew/Cellar/pyyaml/6.0_1: 74 files, 1.5MB
==> Installing meta-package-manager
==> Pouring meta-package-manager--5.13.0.arm64_ventura.bottle.tar.gz
🍺  /opt/homebrew/Cellar/meta-package-manager/5.13.0: 5,885 files, 81.7MB
==> Running `brew cleanup meta-package-manager`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
```

```shell-session
$ mpm --version
mpm, version 5.13.0
{'username': '-', 'guid': '95248f357c293fed4d1ac6cc5ade7ff', 'hostname': '-', 'hostfqdn': '-', 'uname': {'system': 'Darwin', 'node': '-', 'release': '22.4.0', 'version': 'Darwin Kernel Version 22.4.0: Mon Mar  6 20:59:28 PST 2023; root:xnu-8796.101.5~3/RELEASE_ARM64_T6000', 'machine': 'arm64', 'processor': ''}, 'linux_dist_name': '', 'linux_dist_version': '', 'cpu_count': 10, 'fs_encoding': 'utf-8', 'ulimit_soft': 256, 'ulimit_hard': 9223372036854775807, 'cwd': '-', 'umask': '0o2', 'python': {'argv': '-', 'bin': '-', 'version': '3.11.3 (main, Apr 7 2023, 20:13:31) [Clang 14.0.0 (clang-1400.0.29.202)]', 'compiler': 'Clang 14.0.0 (clang-1400.0.29.202)', 'build_date': 'Apr  7 2023 20:13:31', 'version_info': [3, 11, 3, 'final', 0], 'features': {'openssl': 'OpenSSL 1.1.1t  7 Feb 2023', 'expat': 'expat_2.5.0', 'sqlite': '3.41.2', 'tkinter': '', 'zlib': '1.2.11', 'unicode_wide': True, 'readline': True, '64bit': True, 'ipv6': True, 'threading': True, 'urandom': True}}, 'time_utc': '2023-04-10 14:42:29.162105', 'time_utc_offset': 4.0, '_eco_version': '1.0.1'}
```

Thanks everyone for the help! 🤗

**Comment by chenrui333** [comment]
@kdeldycke thanks a lot for your contribution as well!

**Pull request merged**